File: proxy.py
from fastapi import FastAPI, Request, HTTPException
from fastapi.responses import StreamingResponse
from fastapi.middleware.cors import CORSMiddleware
import httpx

app = FastAPI()

# 设置跨域支持
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # 允许所有域名访问，生产环境中可以指定具体域名
    allow_credentials=True,
    allow_methods=["*"],  # 允许所有方法
    allow_headers=["*"],  # 允许所有请求头
)


@app.api_route('/proxy/{full_url:path}', methods=['GET', 'POST', 'PUT', 'DELETE'])
async def proxy(request: Request, full_url: str):
    target_url = full_url if full_url.startswith("http") else f"http://{full_url}"
    timeout = httpx.Timeout(60.0, read=120.0)

    async with httpx.AsyncClient(timeout=timeout) as client:
        try:
            # 处理 GET 请求
            if request.method == "GET":
                response = await client.get(target_url, headers=request.headers, params=request.query_params,
                                            stream=True)

            # 处理 POST 请求
            elif request.method == "POST":
                content_type = request.headers.get('content-type', '')  # 默认空字符串

                # 处理 JSON 数据
                if 'application/json' in content_type:
                    json_data = await request.json()
                    response = await client.post(target_url, json=json_data, headers=request.headers, stream=True)

                # 处理 URL 编码的表单数据
                elif 'application/x-www-form-urlencoded' in content_type:
                    form_data = dict(await request.form())
                    response = await client.post(target_url, data=form_data, headers=request.headers, stream=True)

                # 处理 multipart/form-data 表单数据
                elif 'multipart/form-data' in content_type:
                    form = await request.form()
                    files = {key: (form[key].filename, await form[key].read()) for key in form if
                             isinstance(form[key], UploadFile)}
                    fields = {key: form[key] for key in form if not isinstance(form[key], UploadFile)}
                    response = await client.post(target_url, data=fields, files=files, headers=request.headers,
                                                 stream=True)

                # 如果没有指定 Content-Type，处理默认的 URL 编码表单数据
                elif content_type == '' or 'text/plain' in content_type:
                    form_data = dict(await request.form())
                    response = await client.post(target_url, data=form_data, headers=request.headers, stream=True)

                else:
                    raise HTTPException(status_code=400, detail="Unsupported Content-Type for POST request")

            # 处理 PUT 请求
            elif request.method == "PUT":
                json_data = await request.json()
                response = await client.put(target_url, json=json_data, headers=request.headers, stream=True)

            # 处理 DELETE 请求
            elif request.method == "DELETE":
                response = await client.delete(target_url, headers=request.headers, stream=True)

            else:
                raise HTTPException(status_code=405, detail="Method not allowed")

        except httpx.RequestError as e:
            raise HTTPException(status_code=500, detail=str(e))

    return StreamingResponse(response.aiter_bytes(), status_code=response.status_code, headers=dict(response.headers))


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)

File: dashboard/src/api/modules/proxy.js
/**
 * 模块代理接口
 * 封装 /proxy/:module/* 相关的代理转发接口
 */

import { get, post, put, del } from '../request'
import { API_PATHS } from '../config'

/**
 * 构建代理接口URL
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @returns {string} 完整的代理URL
 */
const buildProxyUrl = (module, path = '') => {
  const basePath = `${API_PATHS.PROXY}/${module}`
  return path ? `${basePath}/${path}` : basePath
}

/**
 * 代理GET请求
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} params - 查询参数
 * @returns {Promise} 代理响应
 */
export const proxyGet = async (module, path = '', params = {}) => {
  const url = buildProxyUrl(module, path)
  return get(url, params)
}

/**
 * 代理POST请求
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} data - 请求数据
 * @returns {Promise} 代理响应
 */
export const proxyPost = async (module, path = '', data = {}) => {
  const url = buildProxyUrl(module, path)
  return post(url, data)
}

/**
 * 代理PUT请求
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} data - 请求数据
 * @returns {Promise} 代理响应
 */
export const proxyPut = async (module, path = '', data = {}) => {
  const url = buildProxyUrl(module, path)
  return put(url, data)
}

/**
 * 代理DELETE请求
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} params - 查询参数
 * @returns {Promise} 代理响应
 */
export const proxyDelete = async (module, path = '', params = {}) => {
  const url = buildProxyUrl(module, path)
  return del(url, params)
}

/**
 * 通用代理请求
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} options - 请求选项
 * @param {string} options.method - 请求方法
 * @param {object} options.params - 查询参数（GET请求）
 * @param {object} options.data - 请求数据（POST/PUT请求）
 * @returns {Promise} 代理响应
 */
export const proxyRequest = async (module, path = '', options = {}) => {
  const { method = 'GET', params, data } = options
  
  switch (method.toUpperCase()) {
    case 'GET':
      return proxyGet(module, path, params)
    case 'POST':
      return proxyPost(module, path, data)
    case 'PUT':
      return proxyPut(module, path, data)
    case 'DELETE':
      return proxyDelete(module, path, params)
    default:
      throw new Error(`不支持的请求方法: ${method}`)
  }
}

/**
 * 代理文件上传
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {FormData} formData - 文件数据
 * @returns {Promise} 上传响应
 */
export const proxyUpload = async (module, path = '', formData) => {
  const url = buildProxyUrl(module, path)
  
  return post(url, formData, {
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  })
}

/**
 * 代理文件下载
 * @param {string} module - 模块名称
 * @param {string} path - 代理路径
 * @param {object} params - 查询参数
 * @returns {Promise} 下载响应
 */
export const proxyDownload = async (module, path = '', params = {}) => {
  const url = buildProxyUrl(module, path)
  
  return get(url, params, {
    responseType: 'blob'
  })
}

// 默认导出所有代理接口
export default {
  proxyGet,
  proxyPost,
  proxyPut,
  proxyDelete,
  proxyRequest,
  proxyUpload,
  proxyDownload
}